File: day51/fourSum.java
/*
18. 四数之和
给定一个包含 n 个整数的数组 nums 和一个目标值 target，判断 nums 中是否存在四个元素 a，b，c 和 d ，使得 a + b + c + d 的值与 target 相等？找出所有满足条件且不重复的四元组。
注意：答案中不可以包含重复的四元组。

来源：力扣（LeetCode）
链接：https://leetcode-cn.com/problems/4sum
 */
 
 class Solution {
    public List<List<Integer>> fourSum(int[] nums, int target) {
        Arrays.sort(nums);
        int size = nums.length;
        List<List<Integer>> res = new LinkedList<>();
        for (int i = 0; i < size - 3; i++) {
            if (i > 0 && nums[i] == nums[i - 1]) continue;

            for (int j = i + 1; j < nums.length - 2; j++) {
                if (j > i + 1 && nums[j] == nums[j - 1]) continue;
                if (nums[i] + nums[j] + nums[j + 1] + nums[j + 2] > target) break;
                if (nums[i] + nums[j] + nums[size - 1] + nums[size - 2] < target) continue;

                int k = j + 1, l = size - 1;
                while (k < l) {
                    int theSum = nums[i] + nums[j] + nums[k] + nums[l];
                    if (theSum < target) k++;
                    else if (theSum > target) l--;
                    else {
                        res.add(Arrays.asList(nums[i], nums[j], nums[k], nums[l]));
                        k++; l--;
                        while (k < l && nums[k] == nums[k - 1]) k++;
                        while (k < l && nums[l] == nums[l + 1]) l--;
                    }
                }
            }
        }
        return res;
        
    }
}

File: day51/fourSum.py
class Solution:
    def fourSum(self, nums: List[int], target: int) -> List[List[int]]:
        nums.sort();
        res = []
        for i in range(len(nums) - 3):
            if (i > 0 and nums[i] == nums[i - 1]): continue
            for j in range(i + 1, len(nums) - 2):
                if (j > i + 1 and nums[j] == nums[j - 1]): continue
                if (nums[i] + nums[j] + nums[-2] + nums[-1]) < target: continue
                if (nums[i] + nums[j] + nums[j + 1] + nums[j + 2] > target): break

                k, l = j + 1, len(nums) - 1
                while k < l:
                    the_sum = nums[i] + nums[j] + nums[k] + nums[l]
                    if the_sum > target: l -= 1
                    elif the_sum < target: k += 1
                    else:
                        res.append([nums[i], nums[j], nums[k], nums[l]])
                        l -= 1; k += 1
                        while k < l and nums[l] == nums[l + 1]: l -= 1
                        while k < l and nums[k] == nums[k - 1]: k += 1
        return res
